### DISCUSSION TITLE
Extend granular related record controls in o2m/m2m interface items

### DISCUSSION THREAD

**Discussion by colemahatmccreary** [directus/directus#15356] [Feature Requests]
I'm imagining an optional button on each item in an o2m/m2m interface, triggering a small dropdown with additional/more granular controls for the relationship and/or related record. 

From a design standpoint it'd be similar to the 'save + stay/save + create new/save as copy' dropdown at the top of the 'edit' view.

Available actions would depend on whether its o2m or m2m, but might include:
- nullifying the FK **or** deleting outright the relationship/"junction" record (for m2m interfaces)
- nullifying the FK **or** deleting outright the related record (o2m **and** m2m interfaces)
- a list of manual flows available on the related and/or "junction" collections

Similar to the 'show a link to the item' option, you'd be able to toggle which of these actions is available for each interface.

For flows triggered in this context, the payload might be modified to include the normal one (from the record containing the flow) **plus** the key and collection of the record from which the interface button was clicked. 

And the deletions/nullifications would of course be constrained by the schema and the current user's permissions.

Thoughts?

**Comment by rijkvanzanten** [comment]
We're currently handling that type of foreign key constraint handling on the field level. Are you suggesting you could change it row-by-row while editing? 🤔

**Reply by colemahatmccreary** [reply]
Yep, i believe you got it.

Instead of (or in addition to) the "X" which triggers field's 'delete' action (which is handled as you say by the field's 'on deselect ...' constraint), you'd basically just make every one of these constraint options available within the dropdown.

Again, you could provide field-level config for which actions to make available on a given interface.

But this way you aren't "limited" to the 'delete' button and a single predefined action. There are situations where, for a single relationship, you may sometimes want to nullify the foreign key and others where the related record should be deleted outright.

Plus this gives the added benefit of accessing other non-delete actions such as flows.

**Reply by rijkvanzanten** [reply]
Totally! Right now, most foreign key constraint handlers are handled on the database level (by setting "ON DELETE CASCADE" etc), so we'll have to do some more investigation to see what options we have in terms of changing it row to row

**Reply by colemahatmccreary** [reply]
I see i see.

It could also be implemented like a set of 'default/built-in flows' available to every interface. This would decouple it from the db level handlers, which **would** still constrain the underlying update/delete queries triggered by the button/flow.

So you essentially have access to a handful of built in processes which attempt a fk update, record delete, etc. (plus the user-defined flows)

**Reply by rijkvanzanten** [reply]
Yeah! That being said, in order to support some newer database types (like planetscale), we have to do some long and hard looking at foreign key constraints to begin with.. A lot of those don't tend to play nicely with foreign key constraints at all, so I think we might have to do a lot more of that magic in the API side of things, at which point we can also do the "per row constraint trigger" type thing 🤔

**Reply by colemahatmccreary** [reply]
Gotcha makes sense – thanks for the insight! 
And great stuff with the recent updates, keep on doing the lord’s work lol

**Reply by rijkvanzanten** [reply]
Thanks! 🙏🏻

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
